### ISSUE TITLE
Error with documentation example using a handler entry point and a router

### ISSUE THREAD

**Issue by vitornesello** [JuliaWeb/HTTP.jl#476]
Julia 1.2.0
HTTP.jl 0.8.8
MbedTLS.jl 0.7.0

Hi guys, I tried to implement this example : https://juliaweb.github.io/HTTP.jl/stable/public_interface/#HTTP.Handlers

The one that uses `JSONHandler` as entry point that calls `HTTP.handle(ANIMAL_ROUTER, ...)` inside.
But I am getting an error because internally HTTP creates a this `HTTP.Handlers.RequestHandlerFunction{typeof(createAnimal)` object which, according to the documentation, is "A function-wrapper type that is a subtype of RequestHandler. Takes a single function as an argument that should be of the form `f(::HTTP.Request) => HTTP.Response`" : https://juliaweb.github.io/HTTP.jl/stable/public_interface/#HTTP.Handlers.RequestHandlerFunction

I really need this kind of functionality, so I would appreciate a lot a fast response and I am willing to help if the fix is not too complicated.

This is the error that I get :

```
┌ Error: error handling request
│   exception =
│    MethodError: no method matching handle(::HTTP.Handlers.RequestHandlerFunction{typeof(createAnimal)}, ::HTTP.Messages.Request, ::Animal)
│    Closest candidates are:
│      handle(::HTTP.Handlers.RequestHandlerFunction, ::HTTP.Messages.Request) at /Users/vitornesello/.julia/packages/HTTP/lZVI1/src/Handlers.jl:253
│      handle(::HTTP.Handlers.Router, ::HTTP.Messages.Request, ::Any...) at /Users/vitornesello/.julia/packages/HTTP/lZVI1/src/Handlers.jl:467
│      handle(::HTTP.Handlers.RequestHandler, ::HTTP.Streams.Stream) at /Users/vitornesello/.julia/packages/HTTP/lZVI1/src/Handlers.jl:273
│      ...
│    Stacktrace:
│     [1] handle(::HTTP.Handlers.Router{Symbol("##371")}, ::HTTP.Messages.Request, ::Animal) at /Users/vitornesello/.julia/packages/HTTP/lZVI1/src/Handlers.jl:467
│     [2] JSONHandler(::HTTP.Messages.Request) at /Users/vitornesello/.julia/dev/AppFramework/src/test.jl:27
│     [3] handle at /Users/vitornesello/.julia/packages/HTTP/lZVI1/src/Handlers.jl:253 [inlined]
│     [4] handle(::HTTP.Handlers.RequestHandlerFunction{typeof(JSONHandler)}, ::HTTP.Streams.Stream{HTTP.Messages.Request,HTTP.ConnectionPool.Transaction{TCPSocket}}) at /Users/vitornesello/.julia/packages/HTTP/lZVI1/src/Handlers.jl:276
│     [5] #4 at /Users/vitornesello/.julia/packages/HTTP/lZVI1/src/Handlers.jl:345 [inlined]
│     [6] macro expansion at /Users/vitornesello/.julia/packages/HTTP/lZVI1/src/Servers.jl:360 [inlined]
│     [7] (::getfield(HTTP.Servers, Symbol("##13#14")){getfield(HTTP.Handlers, Symbol("##4#5")){HTTP.Handlers.RequestHandlerFunction{typeof(JSONHandler)}},HTTP.ConnectionPool.Transaction{TCPSocket},HTTP.Streams.Stream{HTTP.Messages.Request,HTTP.ConnectionPool.Transaction{TCPSocket}}})() at ./task.jl:268
└ @ HTTP.Servers ~/.julia/packages/HTTP/lZVI1/src/Servers.jl:364
ERROR: LoadError: IOError(EOFError() during request(http://localhost:8081/api/zoo/v1/animals))
```

The exact script that I am running :

```julia
using HTTP
using Sockets
using JSON2

mutable struct Animal
    id::Int
    type::String
    name::String
end

const ANIMALS = Dict{Int, Animal}()
const NEXT_ID = Ref(0)
function getNextId()
    id = NEXT_ID[]
    NEXT_ID[] += 1
    return id
end

function JSONHandler(req::HTTP.Request)
    # first check if there's any request body
    body = IOBuffer(HTTP.payload(req))
    if eof(body)
        # no request body
        response_body = HTTP.handle(ANIMAL_ROUTER, req)
    else
        # there's a body, so pass it on to the handler we dispatch to
        response_body = HTTP.handle(ANIMAL_ROUTER, req, JSON2.read(body, Animal))
    end
    return HTTP.Response(200, JSON2.write(response_body))
end

# **simplified** "service" functions
function createAnimal(req::HTTP.Request, animal)
    animal.id = getNextId()
    ANIMALS[animal.id] = animal
    return animal
end

function getAnimal(req::HTTP.Request)
    animalId = HTTP.URIs.splitpath(req.target)[5] # /api/zoo/v1/animals/10, get 10
    return ANIMALS[animalId]
end

function updateAnimal(req::HTTP.Request, animal)
    ANIMALS[animal.id] = animal
    return animal
end

function deleteAnimal(req::HTTP.Request)
    animalId = HTTP.URIs.splitpath(req.target)[5] # /api/zoo/v1/animals/10, get 10
    delete!(ANIMALS, animal.id)
    return ""
end

# define REST endpoints to dispatch to "service" functions
const ANIMAL_ROUTER = HTTP.Router()
HTTP.@register(ANIMAL_ROUTER, "POST", "/api/zoo/v1/animals", createAnimal)
# note the use of `*` to capture the path segment "variable" animal id
HTTP.@register(ANIMAL_ROUTER, "GET", "/api/zoo/v1/animals/*", getAnimal)
HTTP.@register(ANIMAL_ROUTER, "PUT", "/api/zoo/v1/animals", updateAnimal)
HTTP.@register(ANIMAL_ROUTER, "DELETE", "/api/zoo/v1/animals/*", deleteAnimal)

task = @async HTTP.serve(JSONHandler, Sockets.localhost, 8081)

sleep(5.0)

firstanimal = """
{
    "id" : 1,
    "type" : "human",
    "name" : "Vitor"
}
"""

run_response_1 = HTTP.request(
    "POST", "http://localhost:8081/api/zoo/v1/animals",
    [], firstanimal
)
```

**Comment by quinnj** [comment]
Thanks for the report and sorry for the slow response here; I have a fix up here: https://github.com/JuliaWeb/HTTP.jl/pull/480.

**Comment by vitornesello** [comment]
I ended up doing a dirty fix in my side. Thanks a lot though. I will use the official fix

**Comment by mkuegeler** [comment]
Hi, I have installed the latest HTTP package but still get the same error as explained by @vitornesello 
Is  fix #480 available in the latest version of the HTTP package?

**Comment by mattBrzezinski** [comment]
> Hi, I have installed the latest HTTP package but still get the same error as explained by @vitornesello
> Is fix #480 available in the latest version of the HTTP package?

No it's not, the code changes are in master but a new version of `HTTP.jl` has not been tagged yet.

**Comment by mkuegeler** [comment]
ok thanks
